File: merge_subintervals/merge_subintervals.cpp
#include <bits/stdc++.h> 

using namespace std;

vector<vector<int>> mergeIntervals(vector<vector<int>> &intervals)
{
    // Write your code here.
    vector<vector<int>> merged;
    sort(intervals.begin(), intervals.end());
    if(intervals.size()==0) return merged;
    vector<int> temp = intervals[0];
    
    for(auto it: intervals){
        if(it[0]<=temp[1]){
            temp[1] = max(it[1], temp[1]);
        }
        else{
            merged.push_back(temp);
            temp = it;
        }
        
    }
    merged.push_back(temp);
    return merged;
}
int main() {
  vector < vector < int >> arr;
  arr = {{1, 3}, {2, 4}, {2, 6}, {8, 9}, {8, 10}, {9, 11}, {15, 18}, {16, 17}};
  vector < vector < int >> ans = mergeIntervals(arr);

  cout << "Merged Overlapping Intervals are " << endl;

  for (auto it: ans) {
    cout << it[0] << " " << it[1] << "\n";
  }
}

File: merge_subintervals/merge_subintervals.py

from os import *
from sys import *
from collections import *
from math import *

from sys import stdin,setrecursionlimit

class Solution:
    def __init__ (self, start, end):
        self.start = start
        self.end = end

def mergeIntervals(intervals):
    # Write your code here.
    if len(intervals) == 0:
        print("no intervals were given")
        return []
    merged = [intervals[0]]
    k=0
    for i in range(1, len(intervals)):
        if merged[k].end >= intervals[i].start:
            merged[k].end = max(merged[k].end, intervals[i].end)
        else:
            merged.append(intervals[i])
            k+=1
    return merged
    
# n = int(input())
# arr1 = list(map(int, stdin.readline().strip().split(" ")))
# arr2 = list(map(int, stdin.readline().strip().split(" ")))
# arr1.sort()
# arr2.sort()
arr = [[1, 3], [2, 4], [2, 6], [8, 9], [8, 10], [9, 11], [15, 18], [16, 17]]
# arr = []
intervals = []
for i in range(len(arr)):
    a = Solution(arr[i][0], arr[i][1])
    intervals.append(a)

res = mergeIntervals(intervals)

for i in range(len(res)):
    print(res[i].start, res[i].end)


